### ISSUE TITLE
Proposal: Parent-Child selectors

### ISSUE THREAD

**Issue by Andy1605** [angular/angular#7278]
Hi there,

I know the [Angular2 cheat sheet](https://angular.io/cheatsheet) says, that there are no parent-child relationship selectors in Angular2. But i am asking myself: Why?
Angular2 totally needs that (in case i am not missing another super fancy not yet documented way).
It is not only about creating new components it is also about injecting behavior to already existing views which had nothing to do with Angular so far.

Here is an example for an existing Bootstrap-Material view code:

Two toggle buttons:

```
<div class="togglebutton">
   <label>
     <input checked="" type="checkbox"> Toggle button 
   </label>
   <label>
     <input disabled="" type="checkbox">Disabled Toggle button
   </label>
</div>
```

And here two check boxes:

```
<div class="checkbox">
   <label>
     <input checked="" type="checkbox"> Checkbox
   </label>
   <label>
     <input disabled="" type="checkbox">Disabled Checkbox
   </label>
  </div>
```

If you want to inject some custom behavior into those components (checkbox and togglebutton, which behave differently) one would have to select `selector: input[type=checkbox]` in both cases right?
This is where the problem starts: Togglebutton and Checkbox would have the same selector and therefore two controller instances will be created for each `<input type="checkbox" ..>` element. In the constructor of those directives i have to climb up the tree and check what i really need: A checkbox or a togglebutton. But the harm is already done here.

This problem would be simply solved by parent-child selectors:
//for the togglebutton
`selector: 'div.togglebutton > label > input[type=checkbox]'`
//for the checkbox
`selector: 'div.checkbox> label > input[type=checkbox]'`

So Angular2 needs Parent-Child selectors or even better a full compliance to CSS selector syntax.

What do you think? Or am i missing some fancy Angular2 functionality?

**Comment by ericmartinezr** [comment]
See https://github.com/angular/angular/issues/2589 .

**Comment by zoechi** [comment]
What's the problem with adding `togglebutton` to the inputs?

**Comment by Andy1605** [comment]
The problem is, that those are existing views on which i have no access to, since they are coming from a REST service. Since we are rewriting our client we do not want to maintain both frameworks jquery AND angular. We want to reach an unique code base without any inconsistency.

**Comment by zoechi** [comment]
I haven't figured out yet how to use HTML fetched using REST and apply directives or components at all, even though this is an often mentioned feature. Did you get that part already working?  
It doesn't look like Angular2 targets this use case. Angular2 is a lot about building before delivering to the client. In this case you need to control the view anyway and then complex selectors don't add much value.

**Comment by Andy1605** [comment]
Yes of course. It is a simple GET request to a REST api via URL. templateURL wouldn't care much whether i requested an actual HTML file or not. The response is all that counts which is the view markup code.

**Comment by zoechi** [comment]
@Andy1605 Sure it's easy to fetch HTML but do you get any Angular directives and components applied to the dynamically inserted HTML in Angular2?

**Comment by Andy1605** [comment]
I have to admit, that i didn't go that far, yet. Getting the HTML worked, that's where i was pleased for now. ATM i test my views in a separate sandbox project where i play around with Angular2. I didn't doubt that it would work until now, oops.

**Comment by t0f3rchr15** [comment]
**tldr;** Parent-child selectors would provide scope for custom selectors and when selecting native elements.

Another reason for parent-child selectors would be to allow scoping of selectors, especially now that NgModules applies every declaration to every component within itself. For instance: before, I could override table tags with custom components by adding my table directives to the component using the custom table without it bleeding into other components. Now, anywhere a table tag is used in the module, my custom table directives are being applied, causing unexpected behavior. If I could scope my selectors: (i.e. `selector: 'table[my-table] tr'`) I could apply one custom attribute to the table and then be assured that any tr tags which did not reside in `table[my-table]` would not be selected. 

Normally I would just add an attribute to the child tags, but from a plugin perspective, I would rather my users only have to add one attribute to the table tag:
```
<table my-table>
   <thead>
    <tr>
      <th></th>
    </tr>
  </thead>
   ...
```
(preferred)

rather than using:
```
<table my-table>
   <thead my-thead>
    <tr my-tr>
      <th my-th>...</th>
    </tr>
  </thead>
  ...
```
which becomes burdensome to use, especially when there are already multiple input and output directives necessary for the plugins basic function. And since table elements have specific display properties set by the browser, it also doesn't make sense to use custom tag selectors like
```
<my-table>
  <my-row>
   ...
   </my-row>
</my-table>
```

**Comment by sirajc** [comment]
What if you design your component with attribute myTable as selector. And then use @ViewChild to get the thead. @ViewChildren to get all tr and td, and carry out your changes on those elements using APIs provided

**Comment by t0f3rchr15** [comment]
But if I need to add a column of check-boxes to every row, isn't a component is the best way to do that?   And if you create a directive and add your elements through the Renderer inside the table component, won't you will still have a directive that will be applied to any and all tr elements, creating unnecessary bindings?

**Comment by SirZach** [comment]
one year later I'm running into the same request. just showing my support that this shouldn't get closed because it's been open for a long time.

**Comment by fmflame** [comment]
I definitely want this too, I've been researching some time now how can i execute some logic in a directive based on if the element has a certain parent element, but nope... Can't do! :/
